File: rust/byte_freq.rs
use std::io::{Read, Write};

fn main() {
    let mut input = Vec::new();
    std::io::stdin().read_to_end(&mut input).unwrap();

    let mut freq = [0u32; 256];

    for &b in &input {
        freq[b as usize] += 1;
    }

    let mut output = Vec::with_capacity(256 * 5);

    for (j, &count) in freq.iter().enumerate() {
        if count > 0 {
            output.push(j as u8);
            output.extend_from_slice(&count.to_le_bytes());
        }
    }

    std::io::stdout().write_all(&output).unwrap();
}

File: go/byte_freq.go
package main

import (
	"encoding/binary"
	"io"
	"os"
)

func main() {
	input, err := io.ReadAll(os.Stdin)
	if err != nil {
		os.Exit(1)
	}

	var freq [256]uint32
	for _, b := range input {
		freq[b]++
	}

	out := make([]byte, 0, 256*5)
	var tmp [4]byte
	for i, n := range freq {
		if n == 0 {
			continue
		}
		out = append(out, byte(i))
		binary.LittleEndian.PutUint32(tmp[:], n)
		out = append(out, tmp[:]...)
	}

	if _, err := os.Stdout.Write(out); err != nil {
		os.Exit(1)
	}
}
